jwst/clean_flicker_noise/clean_flicker_noise.py

```diff
@@ -0,0 +1,1143 @@
+        The algorithm to use to fit background noise.
+
+    fit_by_channel : bool, optional
+        If set, flicker noise is fit independently for each detector channel.
+        Ignored for MIRI, for subarray data, and for `fit_method` = 'fft'.
+
+    background_method : {'median', 'model', None}, optional
+        If 'median', the preliminary background to remove and restore
+        is a simple median of the background data.  If 'model', the
+        background data is fit with a low-resolution model via
+        `~photutils.background.Background2D`.  If None, the background
+        value is 0.0.
+
+    background_box_size : tuple of int, optional
+        Box size for the data grid used by `Background2D` when
+        `background_method` = 'model'. For best results, use a box size
+        that evenly divides the input image shape.
+
+    background_from_rate : bool, optional
+        If set, and the input data is a ramp model, the background will be
+        fit to the rate images instead of the individual group differences.
+        The preliminary background subtracted from each diff before fitting
+        noise is then rate background * group time.
+
+    mask_science_regions : bool, optional
+        Mask slit/slice regions defined in WCS. Implemented only for
+        NIRSpec science data modes.
+
+    n_sigma : float, optional
+        N-sigma rejection level for finding outliers.
+
+    fit_histogram : bool, optional
+        If set, the 'sigma' used with `n_sigma` for clipping outliers
+        is derived from a Gaussian fit to a histogram of values.
+        Otherwise, a simple iterative sigma clipping is performed.
+
+    single_mask : bool, optional
+        If set, a single mask will be created, regardless of
+        the number of input integrations. Otherwise, the mask will
+        be a 3D cube, with one plane for each integration.
+
+    user_mask : str or None, optional
+        Path to user-supplied mask image.
+
+    save_mask : bool, optional
+        Switch to indicate whether the mask should be saved.
+
+    save_background : bool, optional
+        Switch to indicate whether the fit background should be saved.
+
+    save_noise : bool, optional
+        Switch to indicate whether the fit noise should be saved.
+
+    Returns
+    -------
+    output_model : `~jwst.datamodel.JwstDataModel`
+        Corrected data.
+
+    mask_model : `~jwst.datamodel.JwstDataModel`
+        Pixel mask to be saved or None.
+
+    background_model : `~jwst.datamodel.JwstDataModel`
+        Background model to be saved or None.
+
+    noise_model : `~jwst.datamodel.JwstDataModel`
+        Background model to be saved or None.
+
+    status : {'COMPLETE', 'SKIPPED'}
+        Completion status.  If errors were encountered, status = 'SKIPPED'
+        and the output data matches the input data.  Otherwise,
+        status = 'COMPLETE'.
+    """
+    # Track the completion status, for various failure conditions
```

Comment:
Thanks for pointing that out - I do have a tendency to write monoliths.  I'll see if I can break it up a bit.

Reply:
I refactored a couple of things into helper functions: initial validity checks and intermediate model creation.  I was also able to remove an unnecessary option, which cuts down on the complexity a bit.  The remainder I think would be hard to break down any further without impeding comprehension, but do please let me know if you see somewhere else a refactor might help.  Thanks for the nudge!

Reply:
Long functions are incomprehensible because they include so many steps and details that no one can hold it all in their head.  Most functions should be no more than 30 lines long.  Sometimes due to convenience of computational complexity could require up to 50 lines long.  No function requires more than 100 lines.  Lines of that length have clear and distinct steps with details that are better moved to sub-functions, where the sub-function name is essentially the name of the step being done in the calling function.

For example all of your input checking and preparation based on input parameters that occurs before the line linked below can be handled in separate functions that increases readability of the function, with a series of `_check_*` or `_create_*` sub-functions.

https://github.com/spacetelescope/jwst/blob/0cf53d269ee52fec379b19777f3e374aaeacd248/jwst/clean_flicker_noise/clean_flicker_noise.py#L1002

And since the details of each of those handlers are  now in smaller functions, the details are easier to follow.  I will send you an email with a partial refactor to give you an idea of how this function can be refactored to be smaller and more digestible, with easier to follow sub-functions.

Reply:
I'm not sure that I agree that making functions smaller always makes them easier to follow - I personally prefer not to make helper functions that require a large number of input parameters to do very specific tasks that are not needed anywhere else.  I find the logic easier to follow when it's all in one place.

But I think that's a style issue, and I know you're definitely not the only one who prefers shorter functions! I'll refactor as you suggest.

Reply:
Okay, everything in the main function not directly related to iterating over group images is moved into various helper functions.  I'm glad I wrote thorough unit tests - every line of code is still covered and outputs are the same as before.